Update and fix deps

i-norden opened the pull request.

i-norden committed:
remove gx; use dep instead of govendor; need go 1.12

rmulhol commented on pkg/db/level/storage_trie_reader.go:
```diff
@@ -54,6 +54,7 @@ func (stc *StorageTrieReader) GetStorageTrie(stateTrieLeafNode []byte) (storageT
 			nextStorageHash := storageTrieIterator.Hash()
 			nextStorageNode, err := trieDb.Node(nextStorageHash)
 			if err != nil {
+				println(5)
```
✂️

rmulhol commented on pkg/db/level/storage_trie_reader_test.go:
```diff
@@ -2,49 +2,54 @@ package level_test
+		db.ReturnTrie = mockTrie
+		decoder := mock_rlp.NewMockDecoder()
+		acct := new(state.Account)
+		err := rlp.DecodeBytes(test_helpers.FakeStateLeaf, acct)
```
what's the rationale for deriving `acct` from `rlp.DecodeBytes` rather than setting the return out as `&state.Account{}`? Only asking because it requires the additional import of `go-ethereum/rlp` (in conjunction with `mocks/wrappers/rlp`

i-norden replied:
I changed this because the [old test](https://github.com/vulcanize/eth-block-extractor/blob/master/pkg/db/level/storage_trie_reader_test.go#L26) fails with a 
```
  Expected error:
      <*errors.errorString | 0xc000213e10>: {s: "not found"}
      not found
  not to have occurred
```

When using the empty account and the test_helpers.FakeTrieNode, and from tracing this up this appears to be because there is an empty hash (0x0000000000000000000000000000000000000000000000000000000000000000) being passed to [trieDb.Node](https://github.com/vulcanize/eth-block-extractor/blob/master/pkg/db/level/storage_trie_reader.go#L40) in this case. 

This reminds me, I was a little confused by the [EmptyStorageTrieRoot](https://github.com/vulcanize/eth-block-extractor/blob/master/pkg/db/level/storage_trie_reader.go#L10), how was that value derived?

i-norden replied:
I changed this because the [old test](https://github.com/vulcanize/eth-block-extractor/blob/master/pkg/db/level/storage_trie_reader_test.go#L26) fails with a 
```
  Expected error:
      <*errors.errorString | 0xc000213e10>: {s: "not found"}
      not found
  not to have occurred
```

When using the empty account and the test_helpers.FakeTrieNode, and from tracing this up this appears to be because there is an empty root hash (0x0000000000000000000000000000000000000000000000000000000000000000) being passed to [trieDb.Node](https://github.com/vulcanize/eth-block-extractor/blob/master/pkg/db/level/storage_trie_reader.go#L40) in this case. 

This reminds me, I was a little confused by the [EmptyStorageTrieRoot](https://github.com/vulcanize/eth-block-extractor/blob/master/pkg/db/level/storage_trie_reader.go#L10), how was that value derived?

rmulhol replied:
It's just a `[]byte` version of the empty root hash https://github.com/ethereum/go-ethereum/blob/master/core/state/statedb.go#L43

i-norden replied:
Thank you! Can't yet give you a satisfactory answer as to why the test is failing that way with the empty `&state.Account{}`. This is part of the outdated code and won't be used for the syncAndPublish work, so maybe we ought to extract what we need for that into another repo and forget about trying to update all of this other stuff? Do you have a preference? My main concern with moving what we need to a new repo is preserving your git history.

rmulhol replied:
I'm not too worried about git history since that would be preserved here if you moved to a new repo. That said, I do really dig the idea of getting this repo up to date and working with the latest version of the deps. If what you're saying is that some of the proposed changes are out of scope for the story you're working on, I have no problem zeroing in on the task at hand and pushing those tasks back.

i-norden replied:
No I think you are right, and this is pretty close to being up-to-date. Also I think I am on the cusp of getting ungx to work- will update shortly!

rmulhol commented on pkg/wrappers/trie/database.go:
```diff
@@ -18,5 +18,7 @@ func NewTrieDatabase(db *trie.Database) *Database {
 
 func (td *Database) Node(hash common.Hash) ([]byte, error) {
+	println("HERE")
+	println(hash.Hex())
```
✂️

rmulhol commented on test_helpers/fakes.go:
```diff
@@ -3,12 +3,21 @@ package test_helpers
+		Nonce:    0,
+		Balance:  big.NewInt(10000),
+		CodeHash: common.HexToHash("0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470").Bytes(),
+		Root:     common.HexToHash("0x821e2556a290c86405f8160a2d662042a431ba456b9db265c79bb837c04be5f0"),
```
It looked like some places where this was being decoding, it was ending up as an empty account. Is that just because of mocks? I would have thought that this would have been decoded to an account with these values

i-norden replied:
I think this confusion is coming from this [test](https://github.com/vulcanize/eth-block-extractor/blob/fix_deps/pkg/db/level/storage_trie_reader_test.go#L35), which isn't asserting equality of the [passed](https://github.com/vulcanize/eth-block-extractor/blob/fix_deps/pkg/db/level/storage_trie_reader_test.go#L29) `test_helpers.FakeStateLeaf` and the `&state.Account{}` argument but rather is just [checking](https://github.com/vulcanize/eth-block-extractor/blob/fix_deps/test_helpers/mocks/wrappers/rlp/decode.go#L46) that it is assignable the `&state.Account{}` type.

i-norden replied:
But I might have missed something

rmulhol replied:
🤔 hm yeah, I wonder if it might make sense to add some assertions to that test? especially if we're passing in an account with very specific data

i-norden replied:
Sorry I was wrong, the assertion asserts both the bytes `test_helpers.FakeStateLeaf` and the out type `&state.Account{}` are what were passed to the mock decoder's `Decode` method.

rmulhol commented on test_helpers/mocks/wrappers/core/state/database.go:
```diff
@@ -11,39 +11,31 @@ import (
-		returnDB: nil,
-		trie:     nil,
+		ReturnDB:   nil,
+		ReturnTrie: nil,
```
This is my bad from a long time ago but we probably don't need to specify these values in the initializer if they're just being assigned as `nil`

rmulhol reviewed (commented):
Looking good to m - thanks for doing all this work! Happy to circle back whenever it's stable.

I did want to ask: are there features built into this PR above and beyond making the deps current?

One other random thought, probably best handled in a separate PR - the approach here for computing historical state is pretty out of date and non-performant. Perhaps we want to consider removing that?

Lastly, if we're going to use this in vDB, it might be worth seeing if we can somehow get it work with un-gx-ified dependencies. That might not be possible iirc, but introducing gx is also sort of a pain (to implement but also to maintain) and I'd love to make absolutely certain we need to do that before pulling it in.

i-norden commented:
Thanks for taking a look @rmulhol! 

There are no added features here, just trying to get things working again with the current version of stuff so that we can use some of the packages here as dependencies over at VulcanizeDB for the `syncAndPublish` work. The only packages from here I expect to be using for that stuff are listed in [here](https://github.com/vulcanize/vulcanizedb/blob/syncAndPublish/pkg/ipfs/publisher.go#L25). Part of me is wondering if we should just extract these packages to another repo or something? That would make removing `gx` for those packages a lot simpler.

And I completely whole-heartedly agree about moving away from using `gx` here entirely. I've been toying around with [ungx](https://github.com/karalabe/ungx) today to try to accomplish this, but am running into a bunch of issues with that.

i-norden commented:
Also I can squash this before we merge.

i-norden commented:
Removed gx, and reverted from `govendor` to `dep` since we use that for most our other repos. 

I think we should consider using [go.mod](https://github.com/golang/go/wiki/Modules) but that's another story and from what I can tell it only works with versions of tagged releases and not specific branches or revisions which would be problematic.

rmulhol reviewed (commented):
Looks great!

Maybe worth adding an issue for improving the test in the mock decoder (and perhaps seeing if we can create tests that use only one of `rlp` or `mock_rlp`) if you don't have time to get around to that?

Also idk if you may need to run gx in order to execute the commands in this repo now that the dependencies are un-gxified (which is totally awesome for using this repo as a library btw!), but that may be worth documenting in an issue and/or the readme as well.

Thanks for updating the deps! 😤

i-norden commented:
Moved the decoded account into the `fakes` file so that we lose the `rlp` import in the tests. There isn't any need to run `gx` anymore, I'll update the README to reflect these changes and more. Thanks!!

i-norden committed:
readme changes
